osu.Game/Screens/SelectV2/BeatmapCarouselFilterGrouping.cs

```diff
@@ -114,48 +114,219 @@ void addItem(CarouselItem i)
                         currentGroupItems?.Add(i);
                         currentSetItems?.Add(i);
 
-                        i.IsVisible = i.Model is GroupDefinition || (lastGroup == null && (i.Model is BeatmapSetInfo || currentSetItems == null));
+                        i.IsVisible = i.Model is GroupDefinition || (group == null && (i.Model is BeatmapSetInfo || currentSetItems == null));
                     }
                 }
 
                 return newItems;
             }, cancellationToken).ConfigureAwait(false);
         }
 
-        private GroupDefinition? createGroupIfRequired(FilterCriteria criteria, BeatmapInfo beatmap, GroupDefinition? lastGroup)
+        private List<GroupMapping> getGroups(List<CarouselItem> items, FilterCriteria criteria)
         {
             switch (criteria.Group)
             {
+#pragma warning disable CS0618 // Type or member is obsolete
+                case GroupMode.All:
+#pragma warning restore CS0618 // Type or member is obsolete
+                case GroupMode.NoGrouping:
+                    return new List<GroupMapping> { new GroupMapping(null, items) };
+
                 case GroupMode.Artist:
-                    char groupChar = lastGroup?.Data as char? ?? (char)0;
-                    char beatmapFirstChar = char.ToUpperInvariant(beatmap.Metadata.Artist[0]);
+                    return getGroupsBy(b => defineGroupAlphabetically(b.BeatmapSet!.Metadata.Artist), items);
+
+                case GroupMode.Author:
+                    return getGroupsBy(b => defineGroupAlphabetically(b.BeatmapSet!.Metadata.Author.Username), items);
+
+                case GroupMode.Title:
+                    return getGroupsBy(b => defineGroupAlphabetically(b.BeatmapSet!.Metadata.Title), items);
+
+                case GroupMode.DateAdded:
+                    return getGroupsBy(b => defineGroupByDate(b.BeatmapSet!.DateAdded), items);
+
+                case GroupMode.LastPlayed:
+                    return getGroupsBy(b =>
+                    {
+                        DateTimeOffset? maxLastPlayed = aggregateMax(b, items, bb => bb.LastPlayed);
+
+                        if (maxLastPlayed == null)
+                            return new GroupDefinition(int.MaxValue, "Never");
```

Comment:
This `Order` key (`int.MaxValue`) clashes with the `"Over 5 months ago"` key in `defineGroupByDate()` causing unreliable ordering (`Never` should be after all date categories):
![image](https://github.com/user-attachments/assets/1146cb31-f215-42b3-89b4-0d470b4fe5cd)

Reply:
Should be resolved in https://github.com/ppy/osu/pull/33175/commits/8568ac422fa9b152a31509348671365560768297